src/utilities/indexer/queryFromIndexer.ts

```diff
@@ -0,0 +1,137 @@
+export const QUERY_SIZE = 50;
+
+/** Example Query. */
+const queryBlocks = `
+  query {
+    blocks(orderBy: TIME_STAMP_DESC, first: 3) {
+    totalCount
+      nodes {
+        id
+        timeStamp
+        hash
+      }
+    }
+  }
+`;
+
+export interface FetchedData {
+  data: Record<
+    string,
+    {
+      totalCount?: number;
+      nodes?: Array<Record<string, unknown>>;
+    }
+  >;
+}
+
+export async function queryFromIndexer(query: string = queryBlocks) {
+  logger.debug(
+    `Querying from GraphQL under ${indexer.graphqlEndpoint}, using this payload: ${query} `,
+  );
+  const response = await got
+    .post(indexer.graphqlEndpoint, {
+      json: {
+        query,
+      },
+    })
+    .json<FetchedData>();
+
+  const entities = Object.entries(response.data);
+
+  entities.length > 1 &&
+    logger.error('Please, avoid multiple queries in a single request.');
+
+  const [name, { totalCount, nodes: matches }] = entities[0];
+
+  totalCount ??
+    logger.error(
+      'The query did not ask for total count. Please add field "totalCount" to your query.',
+    );
+
+  matches ??
+    logger.error(
+      'You need to include "nodes" as a field (with subfields) on your query to get matches.',
+    );
+
+  logger.info(
+    `Completed querying '${name}' from GraphQL under ${indexer.graphqlEndpoint}.`,
+  );
+
+  // logger.info('Response from GraphQL: ' + JSON.stringify(response, null, 2));
+
+  logger.info(
+    `Got ${matches ? matches.length : 'none'} out of ${totalCount} '${name}' matching query.`,
+  );
+
+  // matches.forEach((match, index) => {
+  //   logger.info(
+  //     `#${index} match from query: ${JSON.stringify(match, null, 2)}`,
+  //   );
+  // });
+
+  return { totalCount, matches };
+}
+
+export async function* matchesGenerator<T>(
+  queryParams: Parameters<typeof writeQuery>[0],
+): AsyncGenerator<T, void> {
+  if (indexer.graphqlEndpoint === 'NONE') {
+    return;
+  }
+  const writtenQuery = writeQuery(queryParams);
+  const { totalCount, matches } = await queryFromIndexer(writtenQuery);
+
+  if (matches === undefined) {
+    throw new Error(
+      'You need to include "nodes" as a field (with subfields) on your query to get matches.',
+    );
+  }
+  if (totalCount === undefined) {
+    throw new Error(
+      'Please, include "totalCount" as a field on your query for better handling.',
+    );
+  }
+
+  if (totalCount === 0) {
+    logger.debug(
+      `The Indexed Data under "${indexer.graphqlEndpoint}" has no matches for query: ${writtenQuery}.`,
+    );
+    return;
+  }
+
+  if (totalCount === matches.length) {
+    for (const match of matches) {
+      yield match as T;
+    }
+  } else {
+    for (
+      let index = queryParams.offset ?? 0;
+      index < totalCount;
+      index += QUERY_SIZE
+    ) {
+      queryParams.offset = index;
+      const { matches } = await queryFromIndexer(writeQuery(queryParams));
+
+      if (!matches) {
+        continue;
+        // maybe 'break' instead?
+      }
```

Comment:
Under what conditions would it be falsy?

Reply:
If there is nothing on the data base that matches the query.

Reply:
Or if they forget to include "nodes" as a field on their query.

Reply:
Actually only if they forget to include "nodes" to their query. If there is no matches, an empty array will be returned and that is _truthty_. 

I already handle the cases where "nodes" is excluded before the for-loop. So it can't really happen, but the for...of loop still complains that it could happen.

Reply:
Added a comment